From d4ad462992172fa28eb7077ee64a73f42d386c19 Mon Sep 17 00:00:00 2001
From: daiyang5 <daiyang5@xiaomi.com>
Date: Mon, 9 Mar 2026 20:17:49 +0800
Subject: [PATCH 078/223] ANDROID: vendor_hook: Added hooks and a function to
 handle special pages

The following hooks are added to handle special pages:

android_vh_filemap_adjust_folio_flags
  Invoked before adding a folio to the LRU. It allows adjusting the
  active flag to decide whether the folio should be inserted into
  the active or inactive list.

trace_android_vh_folio_skip_activate
  Allows determining whether a folio should bypass the remaining
  steps in activate_locked().

trace_android_vh_folio_trylock_clear_bypass
  Allows determining whether a folio should be added to the
  active list.

Bug: 491161822
Bug: 502721110

Change-Id: Ia204713966fff9609d76dc4f608bc4a2605251c6
Signed-off-by: daiyang5 <daiyang5@xiaomi.com>
(cherry picked from commit 503c1b7b364caddfc06540b1233be6e80152c60e)
---
 drivers/android/vendor_hooks.c |  3 +++
 include/trace/hooks/mm.h       |  4 ++++
 include/trace/hooks/vmscan.h   |  6 ++++++
 mm/filemap.c                   |  1 +
 mm/vmscan.c                    | 14 ++++++++++++--
 5 files changed, 26 insertions(+), 2 deletions(-)

diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index 23c5bd9e7e49..22cc284a66f2 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -607,6 +607,9 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_calculate_totalreserve_pages);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_madvise_cold_pageout_skip);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_rmqueue_pcplist_override_batch);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_do_group_exit);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_folio_skip_activate);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_folio_trylock_clear_bypass);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_filemap_adjust_folio_flags);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_swapmem_gather_init);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_swapmem_gather_add_bypass);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_swapmem_gather_finish);
diff --git a/include/trace/hooks/mm.h b/include/trace/hooks/mm.h
index 729f31fd92b8..1277626cf17c 100644
--- a/include/trace/hooks/mm.h
+++ b/include/trace/hooks/mm.h
@@ -614,6 +614,10 @@ DECLARE_HOOK(android_vh_try_to_unmap_one,
 DECLARE_HOOK(android_vh_mem_cgroup_charge,
 	TP_PROTO(struct folio *folio, struct mem_cgroup **memcg),
 	TP_ARGS(folio, memcg));
+DECLARE_HOOK(android_vh_filemap_adjust_folio_flags,
+	TP_PROTO(struct address_space *mapping, struct folio *folio,
+		pgoff_t index),
+	TP_ARGS(mapping, folio, index));
 DECLARE_HOOK(android_vh_filemap_add_folio,
 	TP_PROTO(struct address_space *mapping, struct folio *folio,
 		pgoff_t index),
diff --git a/include/trace/hooks/vmscan.h b/include/trace/hooks/vmscan.h
index e820aa70e04d..9adf60fb0d05 100644
--- a/include/trace/hooks/vmscan.h
+++ b/include/trace/hooks/vmscan.h
@@ -140,6 +140,12 @@ DECLARE_HOOK(android_vh_direct_reclaim_end,
 DECLARE_HOOK(android_vh_throttle_direct_reclaim_bypass,
 	TP_PROTO(bool *bypass),
 	TP_ARGS(bypass));
+DECLARE_HOOK(android_vh_folio_skip_activate,
+	TP_PROTO(struct folio *folio, bool *skip),
+	TP_ARGS(folio, skip));
+DECLARE_HOOK(android_vh_folio_trylock_clear_bypass,
+	TP_PROTO(struct folio *folio, bool *bypass),
+	TP_ARGS(folio, bypass));
 DECLARE_HOOK(android_vh_shrink_node,
 	TP_PROTO(pg_data_t *pgdat, struct mem_cgroup *memcg),
 	TP_ARGS(pgdat, memcg));
diff --git a/mm/filemap.c b/mm/filemap.c
index 70ff22a1ffad..65ed72600b95 100644
--- a/mm/filemap.c
+++ b/mm/filemap.c
@@ -1031,6 +1031,7 @@ int filemap_add_folio(struct address_space *mapping, struct folio *folio,
 		WARN_ON_ONCE(folio_test_active(folio));
 		if (!(gfp & __GFP_WRITE) && shadow)
 			workingset_refault(folio, shadow);
+		trace_android_vh_filemap_adjust_folio_flags(mapping, folio, index);
 		folio_add_lru(folio);
 	}
 	return ret;
diff --git a/mm/vmscan.c b/mm/vmscan.c
index 9a5b5e91c97d..bdad219d02b0 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -1599,6 +1599,11 @@ static unsigned int shrink_folio_list(struct list_head *folio_list,
 			folio_free_swap(folio);
 		VM_BUG_ON_FOLIO(folio_test_active(folio), folio);
 		if (!folio_test_mlocked(folio)) {
+			bool skip = false;
+
+			trace_android_vh_folio_skip_activate(folio, &skip);
+			if (skip)
+				goto keep_locked;
 			int type = folio_is_file_lru(folio);
 			folio_set_active(folio);
 			stat->nr_activate[type] += nr_pages;
@@ -2246,9 +2251,14 @@ static void shrink_active_list(unsigned long nr_to_scan,
 			 * so we ignore them here.
 			 */
 			if ((vm_flags & VM_EXEC) && folio_is_file_lru(folio)) {
+				bool bypass = false;
+
 				nr_rotated += folio_nr_pages(folio);
-				list_add(&folio->lru, &l_active);
-				continue;
+				trace_android_vh_folio_trylock_clear_bypass(folio, &bypass);
+				if (!bypass) {
+					list_add(&folio->lru, &l_active);
+					continue;
+				}
 			}
 		}
 
-- 
2.50.1

